types/src/proof/proptest_proof.rs

```diff
@@ -28,69 +27,88 @@ fn arb_accumulator_sibling() -> impl Strategy<Value = HashValue> {
     ]
 }
 
-prop_compose! {
-    fn arb_sparse_merkle_proof()(
-        leaf in any::<Option<(HashValue, HashValue)>>(),
-        non_default_siblings in vec(any::<HashValue>(), 0..256usize),
-        total_num_siblings in 0..257usize,
-    ) -> SparseMerkleProof {
-        let mut siblings = non_default_siblings;
-        if !siblings.is_empty() {
-            let total_num_siblings = std::cmp::max(siblings.len(), total_num_siblings);
-            for _ in siblings.len()..total_num_siblings {
-                siblings.insert(0, SPARSE_MERKLE_PLACEHOLDER_HASH.clone());
-            }
-            assert_eq!(siblings.len(), total_num_siblings);
-            (&mut siblings[0..total_num_siblings-1]).shuffle(&mut thread_rng());
-        }
-        SparseMerkleProof::new(leaf, siblings)
-    }
-}
-
-prop_compose! {
-    fn arb_accumulator_consistency_proof()(
-        subtrees in vec(any::<HashValue>(), 0..=MAX_ACCUMULATOR_PROOF_DEPTH),
-    ) -> AccumulatorConsistencyProof {
-        AccumulatorConsistencyProof::new(subtrees)
-    }
+fn arb_non_placeholder_sparse_merkle_sibling() -> impl Strategy<Value = HashValue> {
+    any::<HashValue>().prop_filter("Filter out placeholder sibling.", |x| {
+        *x != *SPARSE_MERKLE_PLACEHOLDER_HASH
+    })
 }
 
-macro_rules! impl_arbitrary_for_proof {
-    ($proof_type: ident, $arb_func: ident) => {
-        impl Arbitrary for $proof_type {
-            type Parameters = ();
-            type Strategy = BoxedStrategy<Self>;
-
-            fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
-                $arb_func().boxed()
-            }
-        }
-    };
+fn arb_sparse_merkle_sibling() -> impl Strategy<Value = HashValue> {
+    prop_oneof![
+        arb_non_placeholder_sparse_merkle_sibling(),
+        Just(*SPARSE_MERKLE_PLACEHOLDER_HASH),
+    ]
 }
 
 impl<H> Arbitrary for AccumulatorProof<H>
 where
-    H: CryptoHasher,
+    H: CryptoHasher + 'static,
 {
     type Parameters = ();
     type Strategy = BoxedStrategy<Self>;
 
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        (0..=MAX_ACCUMULATOR_PROOF_DEPTH)
+            .prop_flat_map(|len| {
+                if len == 0 {
+                    Just(vec![]).boxed()
+                } else {
+                    (
+                        arb_non_placeholder_accumulator_sibling(),
+                        vec(arb_accumulator_sibling(), 0..len),
```

Comment:
`len` is already randomized before hitting this closure, so I think we should just use it here as provided, like `vec(arb_accumulator_sibling(), len - 1)`